store/postgres/src/relational.rs

```diff
@@ -251,6 +252,89 @@ impl Layout {
         Ok(layout)
     }
 
+    pub fn copy_from(
+        &self,
+        logger: &Logger,
+        conn: &PgConnection,
+        base: &Layout,
+        block: EthereumBlockPointer,
+        metadata: &Layout,
+    ) -> Result<(), StoreError> {
+        // This can not be used to copy data to or from the metadata subgraph
+        assert!(!self.subgraph.is_meta());
+        assert!(!base.subgraph.is_meta());
+
+        info!(
+            logger,
+            "Initializing graft by copying data from {} to {}", base.subgraph, self.subgraph
+        );
+
+        // 1. Copy subgraph data
+        // We allow both not copying tables at all from the source, as well
+        // as adding new tables in `self`; we only need to check that tables
+        // that actually need to be copied from the source are compatible
+        // with the corresponding tables in `self`
+        for (dst, src) in self
+            .tables
+            .values()
+            .filter_map(|dst| base.table(&dst.name).map(|src| (dst, src)))
+        {
+            let start = Instant::now();
+            let count = rq::CopyEntityDataQuery::new(dst, src)?.execute(conn)?;
+            info!(logger, "Copied {} {} entities", count, src.object;
+                  "time_ms" => start.elapsed().as_millis());
+        }
+
+        // 2. Copy dynamic data sources and adjust their ID
+        use crate::metadata::dynamic_ethereum_contract_data_source as decds;
+        // Find existing dynamic data sources
+        let start = Instant::now();
+        let dds = decds::table
+            .select(decds::id)
+            .filter(decds::deployment.eq(base.subgraph.as_str()))
+            .load::<String>(conn)?;
+        // Create an equal number of brand new ids
+        let new_dds = (0..dds.len())
+            .map(|_| DynamicEthereumContractDataSourceEntity::make_id())
+            .collect::<Vec<_>>();
+        // Copy the data sources and all their subordinate entities, translating
+        // ids into new ids in the process and attaching them to `self.subgraph`
+        rq::CopyDynamicDataSourceQuery::new(&dds, &new_dds, self.subgraph.as_str())
+            .execute(conn)?;
+        info!(logger, "Copied {} dynamic data sources", dds.len();
+              "time_ms" => start.elapsed().as_millis());
+
+        // 3. Rewind the subgraph. `revert_block` gets rid of everything
+        // including the block passed to it. We want to preserve `block`
+        // and therefore revert `block+1`
+        let start = Instant::now();
+        let revert_to: BlockNumber = (block.number + 1)
+            .try_into()
+            .expect("block numbers fit into an i32");
+        self.revert_block(conn, revert_to)?;
+        metadata.revert_metadata(conn, &self.subgraph, revert_to)?;
```

Comment:
Is `revert_to` the right name if we actually revert to `revert_to - 1`?

Reply:
Hard to come up with a good name here (hence the comment just before this) I changed it to `block_to_revert`

Reply:
That works 😄